File: src/main/java/com/github/morningzeng/toolset/utils/GridBagUtils.java
package com.github.morningzeng.toolset.utils;

import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBPanelWithEmptyText;
import com.intellij.util.ui.GridBag;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.awt.Component;
import java.awt.Container;
import java.awt.GridBagLayout;
import java.util.function.Consumer;

/**
 * @author Morning Zeng
 * @since 2024-05-14
 */
@Data
public class GridBagUtils {

    public static GridBagBuilder<JBPanel<JBPanelWithEmptyText>> builder() {
        final JBPanel<JBPanelWithEmptyText> panel = new JBPanel<>(new GridBagLayout());
        return builder(panel);
    }

    public static <T extends Container> GridBagBuilder<T> builder(final T container) {
        container.setLayout(new GridBagLayout());
        return new GridBagBuilder<T>().container(container);
    }

    @AllArgsConstructor
    public enum GridBagFill {
        NONE(GridBag.NONE),
        BOTH(GridBag.BOTH),
        HORIZONTAL(GridBag.HORIZONTAL),
        VERTICAL(GridBag.VERTICAL),
        ;

        private final int fill;
    }

    static sealed abstract class AbstractGrid<T extends Container> {
        final GridBag bag;

        AbstractGrid(final GridBag bag) {
            this.bag = bag;
        }

        @SuppressWarnings("UnusedReturnValue")
        AbstractGrid<T> fill(GridBagFill fill) {
            this.bag.fill = fill.fill;
            return this;
        }
    }

    @Accessors(fluent = true)
    public static final class GridBagBuilder<T extends Container> extends AbstractGrid<T> {

        @Setter(AccessLevel.PRIVATE)
        private T container;

        GridBagBuilder() {
            super(new GridBag());
        }

        public GridBagBuilder<T> newRow(final Consumer<Row<T>> rowConsumer) {
            this.bag.gridx = 0;
            this.bag.gridy++;
            this.bag.weightx = 0;
            this.bag.gridwidth = 1;
            final Row<T> row = new Row<>(this, this.bag);
            rowConsumer.accept(row);
            return this;
        }

        @Override
        public GridBagBuilder<T> fill(final GridBagFill fill) {
            super.fill(fill);
            return this;
        }

        public T build() {
            return this.container;
        }

    }

    public static final class Row<T extends Container> extends AbstractGrid<T> {

        private final GridBagBuilder<T> builder;

        Row(final GridBagBuilder<T> builder, final GridBag bag) {
            super(bag);
            this.builder = builder;
        }

        @Override
        public Row<T> fill(final GridBagFill fill) {
            super.fill(fill);
            return this;
        }

        public Row<T> gridY(final int gridY) {
            this.bag.gridy = gridY;
            return this;
        }

        public Row<T> gridHeight(final int gridHeight) {
            this.bag.gridheight = gridHeight;
            return this;
        }

        public Cell<T> newCell() {
            this.bag.gridx++;
            this.bag.gridheight = 1;
            return new Cell<>(this, this.bag);
        }

    }

    public static final class Cell<T extends Container> extends AbstractGrid<T> {
        private final Row<T> row;

        Cell(final Row<T> row, final GridBag bag) {
            super(bag);
            this.row = row;
        }

        @Override
        public Cell<T> fill(final GridBagFill fill) {
            super.fill(fill);
            return this;
        }

        public Cell<T> weightX(final double weightX) {
            this.bag.weightx = weightX;
            return this;
        }

        public Cell<T> weightY(final double weightY) {
            this.bag.weighty = weightY;
            return this;
        }

        public Cell<T> gridX(final int gridX) {
            this.bag.gridx = gridX;
            return this;
        }

        public Cell<T> gridWidth(final int gridWidth) {
            this.bag.gridwidth = gridWidth;
            return this;
        }

        public Row<T> add(final Component component) {
            this.row.builder.container.add(component, this.bag);
            return this.row;
        }

    }

}

File: src/main/kotlin/com/github/morningzeng/toolbox/utils/GridBagUtils.kt
package com.github.morningzeng.toolbox.utils

import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBPanelWithEmptyText
import com.intellij.util.ui.GridBag
import java.awt.Component
import java.awt.Container
import java.awt.GridBagLayout

/**
 * @author Morning Zeng
 * @since 2025-05-20
 */
@Suppress("unused")
object GridBagUtils {

    fun builder(): GridBagBuilder<JBPanel<JBPanelWithEmptyText>> = builder(JBPanel<JBPanelWithEmptyText>())

    fun <T : Container> builder(container: T): GridBagBuilder<T> = container.let {
        it.layout = GridBagLayout()
        GridBagBuilder(it)
    }

    enum class GridBagFill(val fill: Int) {
        NONE(GridBag.NONE),
        BOTH(GridBag.BOTH),
        HORIZONTAL(GridBag.HORIZONTAL),
        VERTICAL(GridBag.VERTICAL),
    }

    sealed class AbstractGrid<T : Container>(
        protected var bag: GridBag
    ) {

        open fun fill(fill: GridBagFill): AbstractGrid<T> = this.apply { bag.fill = fill.fill }

    }

    class GridBagBuilder<T : Container>(val container: T) : AbstractGrid<T>(GridBag()) {

        fun build(): T {
            return container
        }

        fun row(consumer: (Row<T>) -> Unit): GridBagBuilder<T> {
            bag.gridx = 0
            bag.gridy++
            bag.weightx = 0.0
            bag.gridwidth = 1
            consumer(Row(this, bag))
            return this
        }

        override fun fill(fill: GridBagFill): GridBagBuilder<T> = this.apply { super.fill(fill) }
    }

    class Row<T : Container>(
        val builder: GridBagBuilder<T>,
        bag: GridBag
    ) : AbstractGrid<T>(bag) {

        override fun fill(fill: GridBagFill): Row<T> = this.apply { super.fill(fill) }

        fun gridY(gridY: Int): Row<T> = this.apply { bag.gridy = gridY }

        fun gridHeight(gridHeight: Int): Row<T> = this.apply { bag.gridheight = gridHeight }

        fun cell(): Cell<T> = this.bag.let {
            it.gridx++
            it.gridheight = 1
            Cell(this, bag)
        }
    }

    class Cell<T : Container>(
        val row: Row<T>,
        bag: GridBag
    ) : AbstractGrid<T>(bag) {

        override fun fill(fill: GridBagFill): Cell<T> = this.apply { super.fill(fill) }

        fun weightX(weightX: Double): Cell<T> = this.apply { bag.weightx = weightX }

        fun weightY(weightY: Double): Cell<T> = this.apply { bag.weighty = weightY }

        fun gridX(gridX: Int): Cell<T> = this.apply { bag.gridx = gridX }

        fun gridWidth(gridWidth: Int): Cell<T> = this.apply { bag.gridwidth = gridWidth }

        fun add(component: Component): Row<T> = this.row.also {
            it.builder.container.add(component, bag)
        }
    }

}